README.md
the plugin allows to change the datasources of the loaded layers. The operation can be performed right clicking on legend items for single layer or globally on a summary table accessible clicking on toolbar plugin icon.


CHANGELOG.txt
V1.0
Release

V2.0 2016/01/29
Datasource browsing feature, Bad layers Handling

V2.1 2016/02/01
Merge PR from https://github.com/czalaviek: New handling button signals - now plugin GUI isn't language-sensitive.

V2.2 2016/04/15
Layer properties not keeped issue fix (native setDataSource method issue)

V2.3 2017/06/19
"Unhandled layers uncaught behaviour" issue fix

V2.4 2017/07/26
"Extent doesn't change applying new datasource to a raster layer" issue fixed

V2.4 2018/09/01
QGIS3 migration; courtesy of Christoph Franke @ChrFr


__init__.py
# -*- coding: utf-8 -*-
"""
/***************************************************************************
 changeDataSource
                                 A QGIS plugin
 right click on layer tree to change layer datasource
                             -------------------
        begin                : 2015-09-29
        copyright            : (C) 2015 by enrico ferreguti
        email                : enricofer@gmail.com
        git sha              : $Format:%H$
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
 This script initializes the plugin, making it known to QGIS.
"""


# noinspection PyPep8Naming
def classFactory(iface):  # pylint: disable=invalid-name
    """Load changeDataSource class from file changeDataSource.

    :param iface: A QGIS interface instance.
    :type iface: QgsInterface
    """
    #
    from .changeDataSource import changeDataSource
    return changeDataSource(iface)


resources_rc.py
# -*- coding: utf-8 -*-

# Resource object code
#
# Created by: The Resource Compiler for PyQt5 (Qt v5.6.2)
#
# WARNING! All changes made in this file will be lost!

from PyQt5 import QtCore

qt_resource_data = b"\
\x00\x00\x05\x99\
\x89\
\x50\x4e\x47\x0d\x0a\x1a\x0a\x00\x00\x00\x0d\x49\x48\x44\x52\x00\
\x00\x00\x20\x00\x00\x00\x20\x08\x06\x00\x00\x00\x73\x7a\x7a\xf4\
\x00\x00\x00\x09\x70\x48\x59\x73\x00\x00\x0b\x13\x00\x00\x0b\x13\
\x01\x00\x9a\x9c\x18\x00\x00\x00\x20\x63\x48\x52\x4d\x00\x00\x6d\
\x75\x00\x00\x73\xa0\x00\x00\xfc\xdd\x00\x00\x83\x64\x00\x00\x70\
\xe8\x00\x00\xec\x68\x00\x00\x30\x3e\x00\x00\x10\x90\xe4\xec\x99\
\xea\x00\x00\x05\x1f\x49\x44\x41\x54\x78\xda\xec\x97\x69\x4c\x94\
\x57\x14\x86\x9f\x6f\x36\x86\x75\x1c\x06\x06\x46\xaa\x80\x82\x80\
\xa3\x54\x16\x43\x4d\xac\xb6\xb4\xd5\x68\x6a\xb5\x44\x6d\x4b\xd4\
\x12\x13\xb1\x26\xb6\x4d\xdc\x5a\x35\xa6\xed\x0f\x6d\xb0\x8b\x8d\
\x8d\x6d\x6c\xeb\x16\xb7\x28\xd5\x6a\x8c\x46\x22\x46\x69\x45\x5c\
\x2b\xb8\x14\x2d\x46\x29\x03\xcc\xb0\x0e\x30\x30\x33\x7c\x33\xcc\
\xed\x0f\x95\x16\x97\x5a\xc5\xc6\xb4\xf5\x24\x37\xb9\x39\x39\xe7\
\xbe\xef\x79\xcf\xcd\x5d\x24\x21\x04\x8f\xd3\x54\xbd\x5d\xa0\x28\
\x7f\x51\x77\x05\x35\x95\xb5\xdd\xfe\xa8\x98\xbe\xdd\xf3\xd1\x53\
\x56\x4a\xf7\xca\x97\x1e\x46\x81\x5b\xa0\x91\x51\x01\xf4\x31\x25\
\xd1\x27\xd4\x1f\x00\x4d\x70\xff\xee\x18\xd9\x51\x45\x4b\xb3\x8b\
\x16\x6b\x39\x00\xb6\x1a\xe7\x5d\xc9\x3c\x30\x81\xa2\xfc\x45\x22\
\x31\x3d\x0d\x63\x74\xc2\x03\xe5\xc9\x8e\x2a\x00\x76\x7d\xb3\x93\
\xec\x85\x5b\xa4\x87\x6e\xc1\x2d\x99\xfd\xe4\x52\xd4\x81\xa1\xa8\
\x82\x13\x7a\x54\x7e\x3b\xa0\xcb\x56\x82\x68\xb7\x20\xd7\x5f\xa4\
\xa2\xc6\x40\x6b\x83\xd4\xfb\x3d\xe0\xef\x2c\x45\x65\xb5\x21\x07\
\x27\x03\xe0\x75\x5c\x01\xc0\xd3\xd1\x8c\x68\xb7\xe0\xb6\xd7\xd3\
\xd9\x56\x87\xbb\xa6\xbc\x47\x5e\x97\x2b\x1a\x88\xee\xe1\x53\xf0\
\x98\xed\xdf\x49\x40\xf6\x74\xfd\xcf\x15\x78\x42\xe0\x09\x81\xff\
\x14\x01\xd5\xd6\x6d\x5b\x08\xd1\x1b\xc9\xcc\x1c\xc3\xd5\x6a\x2b\
\xf6\x76\x77\x5a\xb4\xce\x6f\xbe\xcb\xd1\x11\xfb\x54\x7c\xac\xc3\
\xe3\xf1\x54\xda\xae\x57\x0d\xd4\x19\xfa\x54\x57\xb5\xb8\x57\xfc\
\x23\xef\x01\x01\x08\x21\xfc\x74\xba\x90\x60\xc9\xd7\xb5\x66\x66\
\xee\x1b\x19\xe5\x95\x97\x59\xff\xf1\x26\xbc\x5e\x2f\x6f\x2e\x98\
\x46\x9a\x39\x95\x8d\x9b\x76\x0d\xb0\x05\xea\x50\xa8\x5a\x1e\x1d\
\x01\x59\x96\x09\xd0\x6a\x5e\x68\xb0\xda\xe6\x36\xd6\xd9\xc3\xab\
\xaf\x55\x0f\x29\xad\x29\x25\x28\xbb\x8d\x59\x1f\xe5\xa0\xf2\xd3\
\x10\x98\xd3\xca\x89\xfc\x22\x2a\xcb\x2e\x67\x44\xc6\xbd\x82\x3e\
\xa8\x83\x4e\x36\x3f\x92\xfe\xa9\x86\x0e\x19\x1a\xda\xd2\x64\x5f\
\xfe\xd6\xbc\x9c\x8c\x72\xcb\x65\x02\x09\x40\x37\xba\x03\x63\x0a\
\x34\x07\xd7\x01\x60\x1a\x04\x0a\x9b\x8f\xd9\x4b\x67\xa9\x5b\xbc\
\x76\x52\x62\x86\xb1\x7a\xd9\x64\xf4\xfc\xd4\x7b\x02\xe9\xe9\xc3\
\x45\xd9\xa9\xf3\xde\xd2\x8a\xf3\x78\xc6\xda\xd0\x44\x83\x5e\x0f\
\x2e\x3b\xa8\x23\x6e\xf4\xa7\xa3\x11\xc2\x9e\x85\xd6\xc4\x4b\xb8\
\xad\xf0\x4b\x61\x00\x5d\x9d\x3e\xd0\xf6\x5e\x03\xc5\x85\x0b\x17\
\xda\x0c\xfa\xb0\x93\xc3\xa2\xd2\x51\x7a\xc0\x10\x03\x9d\x5e\x30\
\x68\x4d\x4c\x8e\x99\x41\x56\xcc\x34\xfa\xf8\x85\xe3\xf6\x81\x3e\
\x16\xd4\x1e\x30\x47\x26\xa1\x33\x0e\xc5\xe7\xd7\x1f\x90\xfe\x36\
\x98\xc3\xe9\xbd\x53\x81\x23\x85\x45\x2f\x7f\xba\xe6\x93\x57\x3b\
\xa3\x6b\x89\x4c\x80\x8e\x36\x88\x0d\x89\xe3\xf3\xe7\xf3\x49\x08\
\x1d\x06\xc0\xa5\xc6\xd3\xcc\x3b\xf2\x1a\x35\x1d\xd7\x31\xc4\x41\
\xc9\xb9\x03\x8c\x9f\x6e\x61\xd9\xc2\x0f\x18\x10\x25\xd0\x78\xcb\
\x00\x88\x30\xf4\x5c\xbc\xce\x15\x8d\xc3\xe9\xc5\xe2\x1c\xd8\xed\
\x9b\xb3\xb2\xb0\x07\x63\xd5\xee\xed\xbb\xb3\xea\xb5\x55\xb1\xe6\
\xb9\x20\x75\x42\xab\x03\x72\x86\x2f\xe8\x06\x07\x30\x87\x0d\xe7\
\xf5\xc4\x5c\x56\x9e\x5a\x4a\x60\xb8\x8f\xa8\xb7\xbb\x38\xbb\xa4\
\x94\x1f\x0f\x16\x63\x9e\x3b\x89\x43\x9b\x8f\xdd\xa5\xde\xf0\x7b\
\x82\xf6\xb0\xec\xc9\xd3\x52\x5f\x7c\x6e\xec\xbe\x88\x6c\x44\xca\
\x36\x44\xd2\x7a\x44\x49\xcd\x21\x71\xbb\x15\x5c\xdf\x29\xcc\x1b\
\x35\x22\x79\x13\x22\x62\x2a\x62\x52\xd6\x44\x71\xe6\x6c\xd9\xb9\
\x73\x57\xad\x53\x84\x10\x08\x21\xf8\x72\xe3\x5e\xd1\xd4\x64\x17\
\x8b\xf3\x36\x88\x5b\xbe\xfb\x0d\x45\xee\x9c\x59\x3f\x8f\x4a\x7d\
\x7e\x67\xe3\x79\x90\x94\xe0\x03\x8e\xfc\xb6\xf7\x0e\xa2\xc7\xaa\
\xf7\xe3\x43\xa6\xab\x0b\xda\x2b\x24\xb2\xc7\x4d\x27\x3e\x39\xf9\
\x6b\xa5\x52\x99\xff\xe7\x38\xb7\xdb\xf9\x40\x9b\x50\x2a\xdc\x7f\
\x38\x7e\xc2\xcc\x71\x87\x4d\xef\xca\xfd\xf4\x09\xe0\x71\x80\x02\
\x0d\xb9\xc9\x4b\x98\x9a\x38\x1b\x80\xcd\x97\xbe\x60\xc3\xc5\x55\
\x20\xc9\xa8\x03\xc1\x72\x0a\xc2\x8b\x92\x58\x9d\xb7\x19\x4b\x53\
\x0b\x31\xbe\x82\xbb\xbe\x9c\x5b\x1b\x6c\x74\x99\xdf\xb9\x2b\x70\
\x6d\x5d\x33\x2b\x16\xe5\x48\xaa\x60\x7d\x90\x1c\xa0\x0e\xb6\xd7\
\x6f\x95\x83\x5c\x83\x64\xbd\x71\x7c\x27\xc2\x5f\x66\x75\xe9\x87\
\x6c\x2d\x5f\x85\x00\x9a\x3b\x5b\xd1\xa8\x41\x38\xa1\x76\x8f\x1f\
\xee\x0a\x3f\xfa\xf6\x33\xa1\x0b\x0d\x61\x60\x9c\x89\xca\x92\x02\
\x9e\x19\x33\x12\xaf\xe3\x0a\x72\x5d\x19\x83\xa5\x1f\x00\xf8\xbe\
\x61\x04\x59\x2f\xa5\xa3\xd5\x06\xa0\xd5\x6a\x7a\x10\x58\x9c\xb7\
\xe1\x8f\x8f\xc9\x8e\x6d\x3b\xfa\x1f\x3d\x50\xdc\xef\x57\x9b\xe5\
\x58\x45\xc6\x1e\x4c\x19\xe0\x75\x82\xc7\x73\x23\x58\xa3\x06\xc9\
\x1f\xea\x4b\x60\x50\xf1\x58\x16\xbe\xf7\x3e\xb2\x5a\xc5\xce\x82\
\x13\xd4\x5a\xad\x2c\xcb\x52\x92\x3e\x22\x1c\xb9\xae\x0c\xef\xf5\
\x1b\xe0\x4d\x56\x0f\x47\x2c\xa3\xd8\x65\x79\xfa\x9e\xf2\x17\x6e\
\xff\x4c\x52\x01\x52\x92\x79\x70\x55\x8d\xa5\xb6\xea\x5a\xd5\xd5\
\xbd\xf6\x0a\x26\x36\x9e\x04\xc9\x05\x09\xb9\x20\x3c\x50\xfe\x15\
\xf8\x82\x40\x63\x04\xa7\x68\xb7\x25\xa7\xc5\xe5\xb6\xb8\x14\xfb\
\xc6\x2b\x42\xf1\x0f\x52\x42\xf5\x3a\xb1\x77\x7b\xd9\x4d\xd9\x47\
\xdc\x01\x72\xdf\xcb\x28\x31\x31\x01\xa3\x31\x8c\xe6\xd6\xe6\x83\
\x27\xd7\x1f\x1f\x32\x25\x73\xea\xb7\xd6\x5a\x5b\x78\xf1\xb6\x7d\
\xf3\x51\x40\xa6\xff\xa4\xe5\x86\xb0\xb0\x8e\xc2\x53\xfb\x67\xa4\
\x8f\x4b\xcd\x3b\x73\xfa\xec\x3e\x9d\x31\x1a\x95\x2a\x00\x10\x7f\
\xf9\xf9\xbc\xef\x51\x2c\x84\x40\xad\x51\x4b\x91\x26\x93\x30\x0f\
\x35\x7f\xa7\x75\x05\xee\x4e\x4d\x4f\xa9\xb7\xd5\x5b\x4d\xc7\xd7\
\x1e\x1d\x69\x08\x33\x38\x26\x2d\x98\xb0\xb6\xcd\xee\xb0\x94\x14\
\x1f\x5f\x17\x37\x30\xbe\x51\xa1\x50\xa0\x54\x2a\x6f\xde\xa3\xbd\
\xb3\xdf\x07\x00\xfc\x5f\x52\x4f\x56\x76\xfb\x53\x00\x00\x00\x00\
\x49\x45\x4e\x44\xae\x42\x60\x82\
"

qt_resource_name = b"\
\x00\x07\
\x07\x3b\xe0\xb3\
\x00\x70\
\x00\x6c\x00\x75\x00\x67\x00\x69\x00\x6e\x00\x73\
\x00\x10\
\x0f\x4f\xe3\xd5\
\x00\x63\
\x00\x68\x00\x61\x00\x6e\x00\x67\x00\x65\x00\x44\x00\x61\x00\x74\x00\x61\x00\x53\x00\x6f\x00\x75\x00\x72\x00\x63\x00\x65\
\x00\x08\
\x0a\x61\x5a\xa7\
\x00\x69\
\x00\x63\x00\x6f\x00\x6e\x00\x2e\x00\x70\x00\x6e\x00\x67\
"

qt_resource_struct = b"\
\x00\x00\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00\x01\
\x00\x00\x00\x00\x00\x02\x00\x00\x00\x01\x00\x00\x00\x02\
\x00\x00\x00\x14\x00\x02\x00\x00\x00\x01\x00\x00\x00\x03\
\x00\x00\x00\x3a\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\
"

def qInitResources():
    QtCore.qRegisterResourceData(0x01, qt_resource_struct, qt_resource_name, qt_resource_data)

def qCleanupResources():
    QtCore.qUnregisterResourceData(0x01, qt_resource_struct, qt_resource_name, qt_resource_data)

qInitResources()


browsedatasource.py
# -*- coding: utf-8 -*-

# Form implementation generated from reading ui file 'C:\Users\DEMO\Dropbox\dev\changeDataSource\browsedatasource.ui'
#
# Created: Wed Nov 04 23:39:53 2015
#      by: PyQt4 UI code generator 4.10.2
#
# WARNING! All changes made in this file will be lost!

from builtins import object
from qgis.PyQt import QtCore, QtGui, QtWidgets

try:
    _fromUtf8 = QtCore.QString.fromUtf8
except AttributeError:
    def _fromUtf8(s):
        return s

try:
    _encoding = QtWidgets.QApplication.UnicodeUTF8
    def _translate(context, text, disambig):
        return QtWidgets.QApplication.translate(context, text, disambig, _encoding)
except AttributeError:
    def _translate(context, text, disambig):
        return QtWidgets.QApplication.translate(context, text, disambig)

class Ui_dataSourceBrowser(object):
    def setupUi(self, dataSourceBrowser):
        dataSourceBrowser.setObjectName(_fromUtf8("dataSourceBrowser"))
        dataSourceBrowser.resize(400, 444)
        self.buttonBox = QtWidgets.QDialogButtonBox(dataSourceBrowser)
        self.buttonBox.setGeometry(QtCore.QRect(50, 400, 341, 32))
        self.buttonBox.setOrientation(QtCore.Qt.Horizontal)
        self.buttonBox.setStandardButtons(QtWidgets.QDialogButtonBox.Cancel|QtWidgets.QDialogButtonBox.Ok)
        self.buttonBox.setObjectName(_fromUtf8("buttonBox"))
        self.dataSourceTree = QtWidgets.QTreeView(dataSourceBrowser)
        self.dataSourceTree.setGeometry(QtCore.QRect(10, 11, 381, 381))
        self.dataSourceTree.setObjectName(_fromUtf8("dataSourceTree"))

        self.retranslateUi(dataSourceBrowser)
        self.buttonBox.accepted.connect(dataSourceBrowser.accept)
        self.buttonBox.rejected.connect(dataSourceBrowser.reject)

    def retranslateUi(self, dataSourceBrowser):
        dataSourceBrowser.setWindowTitle(_translate("dataSourceBrowser", "Dialog", None))



ui_changeDSDialog.py
# -*- coding: utf-8 -*-

# Form implementation generated from reading ui file 'Z:\dev\changeDataSource\ui_changeDSDialog.ui'
#
# Created: Tue Sep 29 13:40:52 2015
#      by: PyQt4 UI code generator 4.11.3
#
# WARNING! All changes made in this file will be lost!

from builtins import object
from qgis.PyQt import QtCore, QtGui, QtWidgets

try:
    _fromUtf8 = QtCore.QString.fromUtf8
except AttributeError:
    def _fromUtf8(s):
        return s

try:
    _encoding = QtWidgets.QApplication.UnicodeUTF8
    def _translate(context, text, disambig):
        return QtWidgets.QApplication.translate(context, text, disambig, _encoding)
except AttributeError:
    def _translate(context, text, disambig):
        return QtWidgets.QApplication.translate(context, text, disambig)

class Ui_changeDataSourceDialog(object):
    def setupUi(self, changeDataSourceDialog):
        changeDataSourceDialog.setObjectName(_fromUtf8("changeDataSourceDialog"))
        changeDataSourceDialog.resize(297, 305)
        self.verticalLayout = QtWidgets.QVBoxLayout(changeDataSourceDialog)
        self.verticalLayout.setObjectName(_fromUtf8("verticalLayout"))
        self.label_2 = QtWidgets.QLabel(changeDataSourceDialog)
        self.label_2.setObjectName(_fromUtf8("label_2"))
        self.verticalLayout.addWidget(self.label_2)
        self.selectDatasourceCombo = QtWidgets.QComboBox(changeDataSourceDialog)
        self.selectDatasourceCombo.setObjectName(_fromUtf8("selectDatasourceCombo"))
        self.verticalLayout.addWidget(self.selectDatasourceCombo)
        self.label = QtWidgets.QLabel(changeDataSourceDialog)
        self.label.setObjectName(_fromUtf8("label"))
        self.verticalLayout.addWidget(self.label)
        self.lineEdit = QtWidgets.QPlainTextEdit(changeDataSourceDialog)
        self.lineEdit.setObjectName(_fromUtf8("lineEdit"))
        self.verticalLayout.addWidget(self.lineEdit)
        self.openBrowser = QtWidgets.QPushButton(changeDataSourceDialog)
        self.openBrowser.setObjectName(_fromUtf8("openBrowser"))
        self.verticalLayout.addWidget(self.openBrowser)
        self.buttonBox = QtWidgets.QDialogButtonBox(changeDataSourceDialog)
        self.buttonBox.setOrientation(QtCore.Qt.Horizontal)
        self.buttonBox.setStandardButtons(QtWidgets.QDialogButtonBox.Cancel|QtWidgets.QDialogButtonBox.Ok)
        self.buttonBox.setObjectName(_fromUtf8("buttonBox"))
        self.verticalLayout.addWidget(self.buttonBox)

        self.retranslateUi(changeDataSourceDialog)
        QtCore.QMetaObject.connectSlotsByName(changeDataSourceDialog)

    def retranslateUi(self, changeDataSourceDialog):
        changeDataSourceDialog.setWindowTitle(_translate("changeDataSourceDialog", "undoLayerChanges", None))
        self.label_2.setText(_translate("changeDataSourceDialog", "Datasource Types", None))
        self.label.setText(_translate("changeDataSourceDialog", "URI:", None))
        self.openBrowser.setText(_translate("changeDataSourceDialog", "Browse", None))



changeDataSource_dialog.py
# -*- coding: utf-8 -*-
"""
/***************************************************************************
 changeDataSourceDialog
                                 A QGIS plugin
 right click on layer tree to change layer datasource
                             -------------------
        begin                : 2015-09-29
        git sha              : $Format:%H$
        copyright            : (C) 2015 by enrico ferreguti
        email                : enricofer@gmail.com
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
"""
from __future__ import absolute_import

import os

from qgis.PyQt import QtGui, uic, QtWidgets
from qgis.PyQt.QtCore import pyqtSignal
from qgis.core import QgsBrowserModel, QgsMimeDataUtils
from .changeDataSource_dialog_base import Ui_changeDataSourceDialogBase
from .browsedatasource import Ui_dataSourceBrowser

class changeDataSourceDialog(QtWidgets.QDialog, Ui_changeDataSourceDialogBase):

    def __init__(self, parent=None):
        """Constructor."""
        super(changeDataSourceDialog, self).__init__(parent)
        #QtWidgets.QDialog.__init__(self)
        # Set up the user interface from Designer.
        # After setupUI you can access any designer object by doing
        # self.<objectname>, and you can use autoconnect slots - see
        # http://qt-project.org/doc/qt-4.8/designer-using-a-ui-file.html
        # #widgets-and-dialogs-with-auto-connect
        self.setupUi(self)

    closedDialog = pyqtSignal()

    def closeEvent(self,evnt):
        self.closedDialog.emit()

class dataSourceBrowser(QtWidgets.QDialog, Ui_dataSourceBrowser):

    def __init__(self, parent=None):
        """Constructor."""
        super(dataSourceBrowser, self).__init__(parent)
        #QtWidgets.QDialog.__init__(self)
        # Set up the user interface from Designer.
        # After setupUI you can access any designer object by doing
        # self.<objectname>, and you can use autoconnect slots - see
        # http://qt-project.org/doc/qt-4.8/designer-using-a-ui-file.html
        # #widgets-and-dialogs-with-auto-connect
        self.setupUi(self)
        self.browserModel = QgsBrowserModel()
        self.browserModel.initialize()
        self.dataSourceTree.setModel(self.browserModel)
        self.dataSourceTree.doubleClicked.connect(self.getUriFromBrowser)
        self.dataSourceTree.header().hide()
        self.hide()
        self.buttonBox.accepted.connect(self.acceptedAction)
        self.buttonBox.rejected.connect(self.rejectedAction)
        self.acceptedFlag = None

    def getUriFromBrowser(self,index):
        uriItem = self.browserModel.dataItem(index)
        uri_list = QgsMimeDataUtils.decodeUriList(self.browserModel.mimeData([index]))
        try:
            #print uri_list[0].providerKey,uri_list[0].uri
            self.result =  (uri_list[0].layerType,uri_list[0].providerKey,uri_list[0].uri)
            self.close()
            self.acceptedFlag = True
        except:
            #print "NO VALID URI"
            self.result = (None,None,None)

    def acceptedAction(self):
        self.getUriFromBrowser(self.dataSourceTree.currentIndex())
        self.close()
        self.acceptedFlag = True

    def rejectedAction(self):
        self.close()
        self.acceptedFlag = None

    @staticmethod
    def uri(title=""):
        dialog = dataSourceBrowser()
        dialog.setWindowTitle(title)
        result = dialog.exec_()
        dialog.show()
        if dialog.acceptedFlag:
            return (dialog.result)
        else:
            return (None,None,None)


setdatasource.py
# -*- coding: utf-8 -*-
"""
/***************************************************************************
 undoLayerChangesDialog
                                 A QGIS plugin
 undoLayerChanges
                             -------------------
        begin                : 2014-09-04
        copyright            : (C) 2014 by Enrico Ferreguti
        email                : enricofer@gmail.com
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
"""
from __future__ import print_function
from __future__ import absolute_import

from builtins import str
from builtins import range
from qgis.core import *
from PyQt5.QtCore import *
from PyQt5.QtGui import *
from qgis.PyQt import QtCore, QtGui, QtWidgets
from PyQt5.QtXml import *
from .ui_changeDSDialog import Ui_changeDataSourceDialog
from .changeDataSource_dialog import dataSourceBrowser

from qgis.gui import QgsManageConnectionsDialog, QgsMessageBar
import os.path


class setDataSource(QtWidgets.QDialog, Ui_changeDataSourceDialog):

    def __init__(self,parent):
        QtWidgets.QDialog.__init__(self)
        self.parent = parent
        self.iface = parent.iface
        self.canvas = self.iface.mapCanvas()
        self.setupUi(self)
        self.buttonBox.accepted.connect(self.changeDataSourceAction)
        self.buttonBox.rejected.connect(self.cancelDialog)
        self.selectDatasourceCombo.activated.connect(self.selectDS)
        self.openBrowser.clicked.connect(self.openFileBrowser)
        self.rasterDSList = {"wms":"Web Map Service (WMS)","gdal":"Raster images (GDAL)"}
        self.vectorDSList = {"ogr":"Vector layers (OGR)","delimitedtext": "Delimited Text", "gpx":" GPS eXchange Format", "postgres": "Postgis database layer","spatialite": "Spatialite database layer","oracle": "Oracle spatial database layer","mysql": "Mysql spatial database layer"}
        self.browsable=("ogr","gdal")

    def openFileBrowser(self):
        '''
        method used to open datasource browser dialog to get new provider/uri for the single layer
        '''
        type,provider,fileName = dataSourceBrowser.uri()
        enumLayerTypes = ("vector","raster","plugin")
        if type and enumLayerTypes[self.layer.type()] != type:
            self.iface.messageBar().pushMessage("Error", "Layer type mismatch: %s/%s" % (enumLayerTypes[self.layer.type()],type), level=QgsMessageBar.CRITICAL, duration=4)
        else:
            if fileName:
                self.lineEdit.setPlainText(fileName)
            if provider:
                allSources = [self.selectDatasourceCombo.itemText(i) for i in range(self.selectDatasourceCombo.count())]
                if type in allSources:
                    self.selectDatasourceCombo.setCurrentIndex(allSources.index(provider))
                else:
                    self.selectDatasourceCombo.addItem(provider)
                    self.selectDatasourceCombo.setCurrentIndex(self.selectDatasourceCombo.count()-1)

    def selectDS(self,i):
        '''
        method to catch datasource combo edits. No longer used. Stay here for future uses.
        '''
        pass

    def openDataSourceDialog(self,layer):#,badLayersHandler):
        '''
        method to prep and show single datasource edit dialog
        '''
        self.layer = layer
        #self.badLayersHandler = badLayersHandler
        self.setWindowTitle(layer.name())
        '''
        #if layer is unhandled get unhandled parameters
        if self.parent.badLayersHandler.getActualLayersIds() and self.layer.id() in self.parent.badLayersHandler.getActualLayersIds():

            provider = self.parent.badLayersHandler.getUnhandledLayerFromActualId(self.layer.id())["provider"]
            source = self.parent.badLayersHandler.getUnhandledLayerFromActualId(self.layer.id())["datasource"]
            self.label.setText("unhandled URI:")
        else:
            provider = self.layer.dataProvider().name()
            source = self.layer.source()
            self.label.setText("URI:")
        '''

        provider = self.layer.dataProvider().name()
        source = self.layer.source()
        self.label.setText("URI:")

        if provider == "ogr" or provider == "gdal":
            source = QgsProject.instance().readPath(source)

        if layer.type() == QgsMapLayer.VectorLayer:
            self.populateComboBox(self.selectDatasourceCombo,list(self.vectorDSList.keys()),predef = provider)
        else:
            self.populateComboBox(self.selectDatasourceCombo,list(self.rasterDSList.keys()),predef = provider)
        self.lineEdit.setPlainText(source)
        #self.selectDS(self.selectDatasourceCombo.currentIndex())
        #print source
        self.show()
        self.raise_()
        self.activateWindow()

    def cancelDialog(self):
        '''
        landing method clicking cancel in button box
        '''
        self.hide()

    def exrecoverJoins(self, oldLayer, newLayer):
        '''
        convenience method to rebuild joins if lost
        '''
        for layer in QgsMapLayerRegistry.mapLayers().values():
            if layer.type() == QgsMapLayer.VectorLayer:
                for joinDef in layer.vectorJoins():
                    if joinDef.joinLayerId == oldLayer.id():
                        newJoinDef = joinDef
                        newJoinDef.joinLayerId = newLayer.id()
                        layer.removeJoin(oldLayer.id())
                        layer.addJoin(newJoinDef)

    def changeDataSourceAction(self):
        '''
        landing method clicking apply in button box
        '''
        self.applyDataSource(self.layer,self.selectDatasourceCombo.currentText().lower().replace(' ',''),self.lineEdit.toPlainText())


    def applyDataSource(self,applyLayer,newProvider,newDatasource):
        '''
        method to verify applying datasource/provider before definitive change to avoid qgis crashes
        '''
        self.hide()
        # new layer import
        # fix_print_with_import
        print("applyDataSource", applyLayer.type())
        if applyLayer.type() == QgsMapLayer.VectorLayer:
            # fix_print_with_import
            print("vector")
            probeLayer = QgsVectorLayer(newDatasource,"probe", newProvider)
            extent = None
        else:
            # fix_print_with_import
            print("raster")
            probeLayer = QgsRasterLayer(newDatasource,"probe", newProvider)
            extent = probeLayer.extent()
        if not probeLayer.isValid():
            self.iface.messageBar().pushMessage("Error", "New data source is not valid: "+newProvider+"|"+newDatasource, level=Qgis.Critical, duration=4)
            return None
        #print "geometryTypes",probeLayer.geometryType(), applyLayer.geometryType()

        if applyLayer.type() == QgsMapLayer.VectorLayer and probeLayer.geometryType() != applyLayer.geometryType():
            self.iface.messageBar().pushMessage("Error", "Geometry type mismatch", level=Qgis.Critical, duration=4)
            return None

        newDatasource = probeLayer.source()
        self.setDataSource(applyLayer, newProvider, newDatasource, extent)
        return True

    def setDataSource(self, layer, newProvider, newDatasource, extent=None):
        '''
        Method to write the new datasource to a raster Layer
        '''
        if "setDataSource" in dir(layer):
            qgisVersionOk = True
        else:
            qgisVersionOk = False
            
        XMLDocument = QDomDocument("style")
        XMLMapLayers = XMLDocument.createElement("maplayers")
        XMLMapLayer = XMLDocument.createElement("maplayer")
        context = QgsReadWriteContext()
        layer.writeLayerXml(XMLMapLayer,XMLDocument, context)
        # apply layer definition
        XMLMapLayer.firstChildElement("datasource").firstChild().setNodeValue(newDatasource)
        XMLMapLayer.firstChildElement("provider").firstChild().setNodeValue(newProvider)
        if extent: #if a new extent (for raster) is provided it is applied to the layer
            XMLMapLayerExtent = XMLMapLayer.firstChildElement("extent")
            XMLMapLayerExtent.firstChildElement("xmin").firstChild().setNodeValue(str(extent.xMinimum()))
            XMLMapLayerExtent.firstChildElement("xmax").firstChild().setNodeValue(str(extent.xMaximum()))
            XMLMapLayerExtent.firstChildElement("ymin").firstChild().setNodeValue(str(extent.yMinimum()))
            XMLMapLayerExtent.firstChildElement("ymax").firstChild().setNodeValue(str(extent.yMaximum()))

        XMLMapLayers.appendChild(XMLMapLayer)
        XMLDocument.appendChild(XMLMapLayers)
        layer.readLayerXml(XMLMapLayer, context)
        layer.reload()

        self.iface.actionDraw().trigger()
        self.iface.mapCanvas().refresh()
        self.iface.layerTreeView().refreshLayerSymbology(layer.id())

    def populateComboBox(self,combo,list,dataPayload = None,predef = None,sort = None):
        '''
        procedure to fill specified combobox with provided list
        '''
        combo.blockSignals (True)
        combo.clear()
        model=QStandardItemModel(combo)
        predefInList = None
        for elem in list:
            try:
                item = QStandardItem(str(elem))
            except TypeError:
                item = QStandardItem(str(elem))
            model.appendRow(item)
            if elem == predef:
                predefInList = elem
        if sort:
            model.sort(0)
        combo.setModel(model)
        if predef:
            if predefInList:
                combo.setCurrentIndex(combo.findText(predefInList))
            else:
                combo.insertItem(0,predef)
                combo.setCurrentIndex(0)
        combo.blockSignals (False)
